### PR TITLE
FOUR-6098: Implement debounce for Screen text inputs after user stops typing

### PR DESCRIPTION
**Pull request by danloa**
Needed by the main PR:
https://github.com/ProcessMaker/screen-builder/pull/1240

### CODE REVIEW

File: src/components/mixins/InputDebounce.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    }
+  },
+  methods: {
+    updateInternalValue(event) {
+      /** Rich text areas inputs an event that outputs a clean event string instead
+       * of an event object that has a target and inside a value property as a structure
+       * Checking if the event it is a string, if it is a truthy value, and it doesn't have
+       * the property target in it
+       * */
+      if (window.ProcessMaker && window.ProcessMaker.debounce) {
```
Comment (josechirivella) [inline src/components/mixins/InputDebounce.js:23]: where did the debounce property came from?

File: src/components/mixins/InputDebounce.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        if (typeof event === 'string' && !!event && event.target === undefined) {
+          this.touched = true;
+          return this.updateValue(event);
+        }
+        this.touched = true;
+        this.updateValue(event.target.value);
+      }
+      else {
+        let value = null;
```
Comment (josechirivella) [inline src/components/mixins/InputDebounce.js:32]: all empty variables declarations are null by default
```suggestion
        let value;
```

**Review by josechirivella — CHANGES_REQUESTED** [review]

File: src/components/mixins/InputDebounce.js (hunk 1 above)
Comment (danloa) [reply src/components/mixins/InputDebounce.js]: It is defined in main.js

File: src/components/mixins/InputDebounce.js [hunk 3]
```diff
@@ -20,29 +20,22 @@ export default {
        * Checking if the event it is a string, if it is a truthy value, and it doesn't have
        * the property target in it
        * */
-      if (window.ProcessMaker && window.ProcessMaker.debounce) {
-        if (typeof event === 'string' && !!event && event.target === undefined) {
-          this.touched = true;
-          return this.updateValue(event);
-        }
+      let useDebounce = window.ProcessMaker && (window.ProcessMaker.debounce === undefined || window.ProcessMaker.debounce === true)
```
Comment (josechirivella) [inline src/components/mixins/InputDebounce.js:23]: from our convo, I though we were going to go away from the debounce property inside `window.ProcessMaker`

Comment (josechirivella) [reply src/components/mixins/InputDebounce.js:23]: also, we are mixing types. In some places debounces is being declared as arrays, then here is being declared as boolean type, which one is it

**Review by josechirivella — CHANGES_REQUESTED** [review]

**Review by agustinbusso — APPROVED** [review]
For review notes please refer to https://github.com/ProcessMaker/screen-builder/pull/1240

File: src/components/mixins/InputDebounce.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        this.touched = false;
+        this.$emit('input', valueToEmit);
+        this.$emit('update:value', valueToEmit);
+      }
+    },
+    updateValue: debounce(function (value) {
+      this.touched = false;
+      this.$emit('input', value);
+      this.$emit('update:value', value);
+    }, 500, {'leading': true, 'trainling': false})
```
Comment (josechirivella) [inline src/components/mixins/InputDebounce.js:47]: just an FYI, the correct word is trailing. Is this word being passed to a library component ? if so, we need to fix the typo

**Review by josechirivella — COMMENTED** [review]
Added a comment

File: src/components/mixins/InputDebounce.js (hunk 4 above)
Comment (danloa) [reply src/components/mixins/InputDebounce.js]: Fixed. trailing, leading are  used by lodash. leading=true means to run the debounced function first and await the n milliseconds. By default is leading = false, trailing=true. To improve responsiveness I've found that it was better to run first the function and await.

**Review by josechirivella — APPROVED** [review]

**Review by ryancooley — APPROVED** [review]

**Pull request merged**